### DISCUSSION TITLE
2005/12/07/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93-5-%E4%BD%9C%E4%B8%BA%E5%87%BD%E6%95%B0%E8%BF%94%E5%9B%9E%E5%80%BC%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E4%B8%BA%E4%BB%80%E4%B9%88

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#65] [Announcements]
# 2005/12/07/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93-5-%E4%BD%9C%E4%B8%BA%E5%87%BD%E6%95%B0%E8%BF%94%E5%9B%9E%E5%80%BC%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E4%B8%BA%E4%BB%80%E4%B9%88

Call Library Node 是 LabVIEW 中调用 DLL 函数的节点。如果被调用的函数有一参数数据类型为 char\*，用来输出字符串。我们需要在 CLN 中这个参数对应的左侧接线端连进一个字符串，并且输入字符串的长度要保证大于输出字符串的长度。这个输入字符串的内容是没有用的，它只被用作是被开辟的内存，保存输出字符串。否则，会出现数组越界的运行错误，LabVIEW会莫名其妙死掉。

https://blog.qizhen.xyz/2005/12/07/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93-5-%E4%BD%9C%E4%B8%BA%E5%87%BD%E6%95%B0%E8%BF%94%E5%9B%9E%E5%80%BC%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E4%B8%BA%E4%BB%80%E4%B9%88

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2009年05月23日 04:26 编辑
动态库与LabVIEW之间的数据传递实际上都是复制的。所以一旦数据被传递到LabVIEW中，即便动态库中的数据是局部变量，被销毁，也不会影响到LabVIEW中的数据。

回复
Kun
2009年05月22日 14:45 编辑
如果在动态链接库里面不分配内存的话，怎么能够返回字符串？函数执行完以后，数据由于不是全局的，数据所在内存会被收回的。

回复
Wei
2006年07月22日 16:47 编辑
如果是自己写这个DLL，那就还有一种解决的办法：用LabVIEW的C导出库，用String的Handle传递Char*,在DLL里用的时候，LabVIEW会给这个字符串自动分配内存的，这个问题就不复存在了。
